File: go/mr/coordinator.go
package mr

import (
	"context"
	"log"
	"net"
	"os"
	"strconv"
	sync "sync"
	"time"

	"google.golang.org/grpc"
)

type Coordinator struct {
	UnimplementedWcServer
	mu            sync.Mutex
	mapTasks      map[string]int64
	reduceTasks   map[int]int64
	nWorker       int
	nReduce       int
	timeout       int64
	intermediates map[int][]string
}

func (c *Coordinator) Done() bool {
	c.mu.Lock()
	defer c.mu.Unlock()

	return len(c.mapTasks) == 0 && len(c.reduceTasks) == 0
}

// Cook up a unique-ish UNIX-domain socket name in /var/tmp, for the coordinator.
func CoordinatorSock() string {
	s := "/var/tmp/824-mr-"
	s += strconv.Itoa(os.Getuid())
	return s
}

func MakeCoordinator(files []string, nReduce int) *Coordinator {
	c := Coordinator{nReduce: nReduce}

	c.mapTasks = make(map[string]int64)
	c.intermediates = make(map[int][]string)
	c.reduceTasks = make(map[int]int64)
	c.timeout = 10

	for _, file := range files {
		c.mapTasks[file] = 0
	}
	for i := 0; i < nReduce; i++ {
		c.reduceTasks[i] = 0
	}

	go c.server()
	return &c
}

func (c *Coordinator) server() {
	sockname := CoordinatorSock()
	os.Remove(sockname)
	lis, err := net.Listen("unix", sockname)
	if err != nil {
		log.Fatal("listen error:", err)
	}
	s := grpc.NewServer()
	RegisterWcServer(s, c)
	log.Printf("server listening at %v", lis.Addr())
	if err := s.Serve(lis); err != nil {
		log.Fatalf("failed to server: %v", err)
	}
}

// find pending map task
// if found, sets its start time to time.Now
func (c *Coordinator) NextMapTask() (string, bool) {
	now := time.Now().Unix()
	for task, start := range c.mapTasks {
		if now-start > c.timeout {
			c.mapTasks[task] = now
			return task, true
		}
	}

	return "", false
}

func (c *Coordinator) NextReduceTask() (int, bool) {
	now := time.Now().Unix()
	for task, start := range c.reduceTasks {
		if now-start > c.timeout {
			c.reduceTasks[task] = now
			return task, true
		}
	}

	return -1, false
}

func (c *Coordinator) RequestTask(ctx context.Context, args *RequestArgs) (*RequestReply, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	reply := &RequestReply{
		Kind:    TaskType_None,
		NReduce: int32(c.nReduce),
		Done:    false,
	}

	task, ok := c.NextMapTask()
	if ok {
		c.nWorker++
		reply.Kind = TaskType_Map
		reply.Filename = task
		reply.TaskNo = int32(c.nWorker)
		return reply, nil
	}
	if len(c.mapTasks) == 0 {
		rtask, rok := c.NextReduceTask()
		if rok {
			reply.Kind = TaskType_Reduce
			reply.TaskNo = int32(rtask)
			reply.Intermediates = c.intermediates[rtask]
		}
	}

	reply.Done = len(c.mapTasks) == 0 && len(c.reduceTasks) == 0
	return reply, nil
}

func (c *Coordinator) ReportTask(ctx context.Context, args *ReportArgs) (*ReportReply, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if args.Kind == TaskType_Map {
		delete(c.mapTasks, args.Task)
		log.Printf("Coordinator: removing map task %v, %v remain", args.Task, len(c.mapTasks))
		for i, f := range args.Intermediates {
			c.intermediates[i] = append(c.intermediates[i], f)
		}
	} else {
		delete(c.reduceTasks, int(args.Id))
		log.Printf("Coordinator: removing reduce task %v, %v remain", args.Id, len(c.reduceTasks))
	}
	return &ReportReply{}, nil
}

File: rust/mr/src/mr/coordinator.rs
use std::sync::{Mutex, Arc};
use std::{collections::HashMap, time::SystemTime};
use tonic::{Request, Response, Status};

use crate::mapreduce::map_reduce_server::MapReduce;
use crate::mapreduce::{ReportArgs, ReportReply, RequestArgs, RequestReply, TaskType};

#[derive(Debug)]
pub struct Coordinator {
    map_tasks: HashMap<String, SystemTime>,
    reduce_tasks: HashMap<usize, SystemTime>,
    n_worker: usize,
    n_reduce: usize,
    timeout: u64,
    intermediates: HashMap<usize, Vec<String>>,
}

impl Coordinator {
    pub fn new(timeout: u64, n_reduce: usize, files: Vec<String>) -> Self {
        let mut c  = Self {
            map_tasks: HashMap::new(),
            reduce_tasks: HashMap::new(),
            intermediates: HashMap::new(),
            n_worker: 0,
            n_reduce,
            timeout,
        };
        for f in files {
            c.map_tasks.insert(f, SystemTime::UNIX_EPOCH);
        }
        for i in 0..n_reduce {
            c.reduce_tasks.insert(i, SystemTime::UNIX_EPOCH);
        }
        c
    }

    // true when no map tasks or reduce tasks remain
    pub fn done(&self) -> bool {
        self.map_tasks.is_empty() && self.reduce_tasks.is_empty()
    }

    // Find pending map task. If found, set its start time
    pub fn next_map_task(&mut self) -> Option<String> {
        let now = SystemTime::now();
        for (task, start) in &mut self.map_tasks {
            if now.duration_since(*start).unwrap().as_secs() > self.timeout {
                *start = now;
                return Some(String::from(task));
            }
        }
        None
    }

    // Find pending reduce task and set its start time
    pub fn next_reduce_task(&mut self) -> Option<usize> {
        let now = SystemTime::now();
        for (task, start) in &mut self.reduce_tasks {
            if now.duration_since(*start).unwrap().as_secs() > self.timeout {
                *start = now;
                return Some(*task);
            }
        }
        None
    }
}

#[derive(Debug)]
pub struct CoordinatorService {
    pub coordinator: Arc<Mutex<Coordinator>>,
}

#[tonic::async_trait]
impl MapReduce for CoordinatorService {
    // Called by workers to request a new task
    async fn request_task(
        &self,
        _args: Request<RequestArgs>,
    ) -> Result<Response<RequestReply>, Status> {
        let mut c = self.coordinator.lock().unwrap();
        let mut reply = RequestReply {
            n_reduce: c.n_reduce as i32,
            ..Default::default()
        };

        if let Some(task) = c.next_map_task() {
            c.n_worker += 1;
            reply.kind = TaskType::Map as i32;
            reply.filename = task;
            reply.task_no = c.n_worker as i32;
            return Ok(Response::new(reply));
        }

        if c.map_tasks.is_empty() {
            if let Some(task) = c.next_reduce_task() {
                reply.kind = TaskType::Reduce as i32;
                reply.task_no = task as i32;
                reply.intermediates = c.intermediates[&task].clone();
            }
        }

        reply.done = c.done();
        Ok(Response::new(reply))
    }

    // Called by workers to report task results
    async fn report_task(
        &self,
        args: Request<ReportArgs>,
    ) -> Result<Response<ReportReply>, Status> {
        let mut c = self.coordinator.lock().unwrap();

        let args = args.get_ref();
        if args.kind == TaskType::Map as i32 {
            c.map_tasks.remove(&args.task);
            println!(
                "Coordinator: finished map task {}: {} to go",
                args.task,
                c.map_tasks.len()
            );
            for (i, f) in args.intermediates.iter().enumerate() {
                c.intermediates
                    .entry(i)
                    .or_insert(Vec::new())
                    .push(String::from(f));
            }
        } else {
            c.reduce_tasks.remove(&(args.id as usize));
            println!(
                "Coordinator: finished reduce task {}: {} to go",
                args.task,
                c.reduce_tasks.len()
            );
        }

        Ok(Response::new(ReportReply::default()))
    }
}

impl CoordinatorService {
    pub fn new(timeout: u64, n_reduce: usize, files: Vec<String>) -> Self {
        Self {
            coordinator: Arc::new(Mutex::new(Coordinator::new(timeout, n_reduce, files))),
        }
    }
}
